sql/sql_acl.cc

```diff
@@ -13145,7 +13143,67 @@ bool Sql_cmd_grant_object::grant_stage0_exact_object(THD *thd,
 }
 
 
-bool Sql_cmd_grant_table::execute_exact_table(THD *thd, TABLE_LIST *table)
+bool Sql_cmd_grant_table::execute_grant_database(THD *thd)
+{
+  Grant_tables tables;
+  bool result;
+  const bool no_auto_create_users= MY_TEST(thd->variables.sql_mode &
+                                           MODE_NO_AUTO_CREATE_USER);
+  char tmp_db[SAFE_NAME_LEN+1];
+  const char *db= m_gp.db().str;
+  /* Convert database name to lower case if lower_case_table_names. */
+  if (lower_case_table_names)
+  {
+    char *end= strnmov(tmp_db, db, sizeof(tmp_db));
+    if (end >= tmp_db + sizeof(tmp_db))
+    {
+      my_error(ER_WRONG_DB_NAME ,MYF(0), db);
+      return true;
+    }
+    my_casedn_str(files_charset_info, tmp_db);
+    db= tmp_db;
+  }
+
+  /* Only allow DB level grants. */
+  if ((m_gp.object_privilege() & DB_ACLS) != m_gp.object_privilege())
+  {
+    my_error(ER_WRONG_USAGE, MYF(0), "DB GRANT", "GLOBAL PRIVILEGES");
+    return true;
+  }
+
+  WSREP_TO_ISOLATION_BEGIN(WSREP_MYSQL_DB, NULL, NULL);
+
+  if (tables.open_and_lock(thd, Table_user | Table_db, TL_WRITE))
+    return true;
+
+  mysql_rwlock_wrlock(&LOCK_grant);
+  mysql_mutex_lock(&acl_cache->lock);
+  grant_version++;
+
+  result= mysql_grant_db(thd, tables,
+                         m_resolved_users, db,
+                         {m_gp.object_privilege(), is_revoke()},
+                         m_create_new_users, no_auto_create_users);
+
+  mysql_mutex_unlock(&acl_cache->lock);
+  /* Write to binlog if statement succeeded. */
+  if (!result)
+    result= write_bin_log(thd, TRUE, thd->query(), thd->query_length());
+  mysql_rwlock_unlock(&LOCK_grant);
+
+  if (!result)
+    my_ok(thd);
```

Comment:
why not to put this tail into the caller? now you have it duplicated.
and the head is pretty much the same, except for `lower_case_table_names` part. you can move the head into the caller too